attestation/k8smanifest/k8smanifest.go

```diff
@@ -0,0 +1,485 @@
+			nameVal = nm
+		}
+	}
+
+	baseKey := fmt.Sprintf("k8smanifest:%s:%s:%s", filePath, kindVal, nameVal)
+	subjectKey := baseKey
+	suffix := 1
+	for {
+		_, loaded := a.subjectDigests.Load(subjectKey)
+		if !loaded {
+			break
+		}
+		suffix++
+		subjectKey = fmt.Sprintf("%s#%d", baseKey, suffix)
+	}
+
+	ro := RecordedObject{
+		FilePath:      filePath,
+		Kind:          kindVal,
+		Name:          nameVal,
+		Data:          finalObj,
+		CanonicalJSON: canonStr,
+		SubjectKey:    subjectKey,
+	}
+
+	return []byte(canonStr), ro, nil
+}
+
+// runDryRun executes kubectl apply --dry-run=server -o json -f -
+func (a *Attestor) runDryRun(doc map[string]interface{}) (map[string]interface{}, error) {
+	y, err := yaml.Marshal(doc)
+	if err != nil {
+		return nil, err
+	}
+	args := []string{"apply", "--dry-run=server", "-o", "json", "-f", "-"}
+	if a.KubeconfigPath != "" {
+		args = append(args, "--kubeconfig", a.KubeconfigPath)
+	}
+	cmd := exec.Command("kubectl", args...)
+	cmd.Stdin = bytes.NewReader(y)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("kubectl dry-run error: %s (output=%q)", err, string(out))
+	}
+	var outMap map[string]interface{}
+	if e := json.Unmarshal(out, &outMap); e != nil {
+		return nil, fmt.Errorf("unmarshal after dry-run: %w", e)
+	}
+	return outMap, nil
+}
+
+// removeEphemeralFields removes ephemeral fields & annotations
+func (a *Attestor) removeEphemeralFields(obj map[string]interface{}) {
+	for _, ef := range a.ephemeralFields {
+		removeNested(obj, ef)
+	}
+	removeEphemeralAnnotations(obj, a.ephemeralAnnotations)
+}
+
+// removeNested handles dot-separated paths, e.g. "metadata.name"
+func removeNested(obj map[string]interface{}, path string) {
+	parts := strings.Split(path, ".")
+	cur := obj
+	for i := 0; i < len(parts)-1; i++ {
+		sub, ok := cur[parts[i]].(map[string]interface{})
+		if !ok {
+			return
+		}
+		cur = sub
+	}
+	delete(cur, parts[len(parts)-1])
+}
+
+// removeEphemeralAnnotations removes ephemeral annotation keys
+func removeEphemeralAnnotations(obj map[string]interface{}, ephemeralKeys []string) {
+	md, _ := obj["metadata"].(map[string]interface{})
+	if md == nil {
+		return
+	}
+	ann, _ := md["annotations"].(map[string]interface{})
+	if ann == nil {
+		return
+	}
+	for _, k := range ephemeralKeys {
+		delete(ann, k)
+	}
+}
+
+// Subjects returns computed subject digests
+func (a *Attestor) Subjects() map[string]cryptoutil.DigestSet {
+	out := make(map[string]cryptoutil.DigestSet)
+	a.subjectDigests.Range(func(k, v interface{}) bool {
+		key := k.(string)
+		ds := v.(cryptoutil.DigestSet)
+		out[key] = ds
+		return true
+	})
+	return out
+}
+
+// toCanonicalJSON performs stable JSON encoding
+func toCanonicalJSON(v interface{}) (string, error) {
+	var sb strings.Builder
+	if err := encodeCanonical(v, &sb); err != nil {
+		return "", err
+	}
+	return sb.String(), nil
+}
+
+func encodeCanonical(val interface{}, sb *strings.Builder) error {
```

Comment:
we probably want a common function to use here with this and the JSON attestor

Reply:
We could stick it in `/pkg`?

Reply:
wait don't we have https://github.com/cyberphone/json-canonicalization?